scripts/question_answering/train_question_answering.py

```diff
@@ -0,0 +1,769 @@
+
+                if eval_result['f1'] > max_dev_f1:
+                    max_dev_f1 = eval_result['f1']
+                    max_dev_exact = eval_result['exact_match']
+                    max_iteration = iteration
+                    early_stop_tries = 0
+                else:
+                    early_stop_tries += 1
+                    if early_stop_tries < options.early_stop:
+                        print('Results decreased for {} times'.format(early_stop_tries))
+                    else:
+                        print('Results decreased for {} times. Stop training. '
+                              'Best results are stored at {} params file. F1={}, EM={}' \
+                              .format(options.early_stop + 1, max_iteration,
+                                      max_dev_f1, max_dev_exact))
+                        break
+
+            for l in losses:
+                avg_loss += l.mean().as_in_context(avg_loss.context)
+
+            iteration += 1
+
+        mx.nd.waitall()
+
+        avg_loss /= (i * len(ctx))
+
+        # block the call here to get correct Time per epoch
+        avg_loss_scalar = avg_loss.asscalar()
+        epoch_time = time() - e_start
+
+        print('\tEPOCH {:2}: train loss {:6.4f} | batch {:4} | lr {:5.3f} '
+              '| throughtput {:5.3f} of samples/sec | Time per epoch {:5.2f} seconds'
+              .format(e, avg_loss_scalar, options.batch_size, trainer.learning_rate,
+                      records_per_epoch_count / epoch_time, epoch_time))
+
+        save_model_parameters(net.collect_params() if ema is None else ema.get_params(), e, options)
+        save_trainer_parameters(trainer, e, options)
+
+        if options.terminate_training_on_reaching_F1_threshold:
+            evaluate_options = copy.deepcopy(options)
+            evaluate_options.epochs = e
+            eval_result = run_evaluate_mode(evaluate_options, net, ema)
+
+            if eval_result['f1'] >= options.terminate_training_on_reaching_F1_threshold:
+                print('Finishing training on {} epoch, because dev F1 score is >= required {}. {}'
+                      .format(e, options.terminate_training_on_reaching_F1_threshold, eval_result))
+                break
+
+    print('Training time {:6.2f} seconds'.format(time() - train_start))
+
+
+def warm_up_steps(iteration, options):
+    """Returns learning rate based on current iteration. Used to implement learning rate warm up
+    technique
+
+    Parameters
+    ----------
+    iteration : `int`
+        Number of iteration
+    options : `Namespace`
+        Training options
+
+    Returns
+    -------
+    learning_rate : float
+        Learning rate
+    """
+    if options.resume_training:
+        return options.lr
+
+    return min(options.lr, options.lr * (math.log(iteration) / math.log(options.lr_warmup_steps)))
+
+
+def get_gradients(model, ctx, options):
+    """Get gradients and apply gradient decay to all layers if required.
```

Comment:
Is this different to the weight decay provided by the mxnet optimizer?

Reply:
A little bit. This particular method applies it only to UNK token of the embedding.